### PR TITLE
feat(slippage): small order slippage v2

### PR DESCRIPTION
**Pull request by alfetopito**
# Update 2024/10/09

Cap now at 5%
When suggested slippage > 2%, show warning:
image
image

Text is obviously not final.

# Update 2024/10/03

Added feature flag `smartSlippageFeeMultiplierPercentage`, that returns the multiplier value and can still be used to feature flag the individual feature of small orders slippage.

Also, slippage is now capped at 50%.

## New testing steps

1. Enable `smartSlippageFeeMultiplierPercentage` and disable `isSmartSlippageEnabled`
2. Pick sell/buy/amounts and observe the suggested slippage
* Should still work independent from BFF's slippage
3. Pick a tiny sell amount and observe the UI and the console logs
* Max suggested UI value should be 50%
* Console logs should show more than `5000`
image

One more fun test is to add/pick new values to be served when on for the new feature flag:
image
And then observe the slippage change accordingly in the UI.

# Summary

Supersedes https://github.com/cowprotocol/cowswap/pull/4911

Instead of using the fee % in relation to order size, calculate how much slippage is needed if the fee is 50% higher.
This value is added to the one returned by BFF.

⚠️ Tooltips still not updated

# To Test

1. Make sure the Launch darkly flag is on
4. This will depend on current network conditions, so play around with input amounts to get it within those ranges
5. Place the order
* Should trade somewhat fast, hopefully. Not guaranteed, though.

You can also observe all the smart slippage values by filtering the console logs by `SmartSlippage`
- Granular value proposed on #4911 
- The 50% fee increased from this PR
- BFF suggested slippage amount
image

### CODE REVIEW

File: apps/cowswap-frontend/src/modules/swap/updaters/SmartSlippageUpdater.ts [hunk 1]
```diff
@@ -42,12 +44,115 @@ export function SmartSlippageUpdater() {
...
+      // (sellAmount + feeAmount * 1.5) / (sellAmount + feeAmount) - 1
+      // (inputAmountWithFee + feeAsCurrency * 1.5) / inputAmountWithFee - 1
+      return (
+        inputAmountWithFee
+          .add(feeAsCurrency.multiply(FEE_MULTIPLIER_FACTOR))
+          // !!! Need to convert to fraction before division to not lose precision
+          .asFraction.divide(inputAmountWithFee.asFraction)
+          .subtract(ONE)
+      )
+    }
```
Comment (alfetopito) [inline apps/cowswap-frontend/src/modules/swap/updaters/SmartSlippageUpdater.ts:115]: Core logic for calculating slippage based on fee

File: apps/cowswap-frontend/src/common/updaters/FeesUpdater.ts [hunk 2]
```diff
@@ -61,27 +61,60 @@ function quoteUsingSameParameters(currentParams: FeeQuoteParams, quoteInfo: Quot
...
     hasSameReceiver
   // 2 checks: if there's a quoteInfo user address (meaning quote was already calculated once) and one without
   // in case user is not connected
   return userAddress ? currentUserAddress === userAddress && paramsWithoutAddress : paramsWithoutAddress
 }
 
+/**
+ * Compares appData without taking into account the `quote` metadata
+ */
+function compareAppDataWithoutQuoteData<T extends string | undefined>(a: T, b: T): boolean {
```
Comment (shoom3301) [inline apps/cowswap-frontend/src/common/updaters/FeesUpdater.ts:82]: Just to keep the file clean, would be better to move the new functions into a new file

File: apps/cowswap-frontend/src/legacy/state/price/hooks.ts [hunk 3]
```diff
@@ -71,7 +71,7 @@ export const useGetQuoteAndStatus = (params: QuoteParams): UseGetQuoteAndStatus
   const isGettingNewQuote = Boolean(isLoading && !quote?.price?.amount)
   const isRefreshingQuote = Boolean(isLoading && quote?.price?.amount)
 
-  return { quote, isGettingNewQuote, isRefreshingQuote }
+  return useMemo(() => ({ quote, isGettingNewQuote, isRefreshingQuote }), [quote, isGettingNewQuote, isRefreshingQuote])
```
Comment (shoom3301) [inline apps/cowswap-frontend/src/legacy/state/price/hooks.ts:74]:
```suggestion
  return useSafeMemoObject({ quote, isGettingNewQuote, isRefreshingQuote })
```

File: apps/cowswap-frontend/src/modules/swap/pure/Row/RowSlippageContent/index.tsx [hunk 4]
```diff
@@ -137,11 +144,11 @@ export function RowSlippageContent(props: RowSlippageContentProps) {
       <TextWrapper textAlign="right">
         {showSettingOnClick ? (
           <ClickableText onClick={toggleSettings}>
-            {displaySlippage}{displayDefaultSlippage}
+            {loading ? loading : (<>{displaySlippage}{displayDefaultSlippage}</>)}
```
Comment (shoom3301) [inline apps/cowswap-frontend/src/modules/swap/pure/Row/RowSlippageContent/index.tsx:147]: The code seems to be duplicated bellow, can it be cached in a const?

File: apps/cowswap-frontend/src/modules/swap/updaters/SmartSlippageUpdater.ts [hunk 5]
```diff
@@ -42,12 +44,115 @@ export function SmartSlippageUpdater() {
...
+    const slippage = tradeSizeSlippageBps + (bffSlippageBps || 0)
+
+    setSmartSwapSlippage(slippage)
+  }, [bffSlippageBps, setSmartSwapSlippage, tradeSizeSlippageBps])
+
+  // TODO: remove before merging
   useEffect(() => {
-    setSmartSwapSlippage(typeof slippageBps === 'number' ? slippageBps : null)
-  }, [slippageBps, setSmartSwapSlippage])
+    console.log(`SmartSlippageUpdater`, {
```
Comment (shoom3301) [inline apps/cowswap-frontend/src/modules/swap/updaters/SmartSlippageUpdater.ts:63]: console.log

File: apps/cowswap-frontend/src/modules/swap/updaters/SmartSlippageUpdater.ts [hunk 6]
```diff
@@ -42,12 +44,115 @@ export function SmartSlippageUpdater() {
...
+const ZERO = new Fraction(0)
+
+/**
+ * Calculates smart slippage in bps, based on quoted fee
+ *
+ * Apply a multiplying factor to the fee (e.g.: 50%), and from there calculate how much slippage would be needed
+ * for the limit price to take this much more fee.
+ * More relevant for small orders in relation to fee amount, negligent for larger orders.
+ */
+function useSmartSlippageFromFeePercentageV2(): number {
```
Comment (shoom3301) [inline apps/cowswap-frontend/src/modules/swap/updaters/SmartSlippageUpdater.ts:84]: The hooks deserves its own file :) 
I would also suggest to extract the whole logic into a pure function and cover it by unit-tests

File: apps/cowswap-frontend/src/modules/swap/updaters/SmartSlippageUpdater.ts [hunk 7]
```diff
@@ -42,12 +44,115 @@ export function SmartSlippageUpdater() {
...
+  const { trade } = useDerivedSwapInfo() || {}
```
Comment (shoom3301) [inline apps/cowswap-frontend/src/modules/swap/updaters/SmartSlippageUpdater.ts:85]: I'm not sure about that, but if possible, would be better to replace `useDerivedSwapInfo` usage with `useReceiveAmountInfo`. `useDerivedSwapInfo` is a bit legacy.

Comment (shoom3301) [reply apps/cowswap-frontend/src/modules/swap/updaters/SmartSlippageUpdater.ts:85]: Also, `useDerivedSwapInfo` is a wrapper on top of `getQuoteAmountsAndCosts()` function from cow-sdk, so you could even move your fee logic in the SDK.

File: apps/cowswap-frontend/src/modules/swap/updaters/SmartSlippageUpdater.ts (hunk 1 above)
Comment (shoom3301) [reply apps/cowswap-frontend/src/modules/swap/updaters/SmartSlippageUpdater.ts:115]: The logic seems to match [the doc](https://docs.google.com/spreadsheets/d/1WtlJ8qKh-DlZy8xS7pyXZkd3MdXysT5J2T4aUHhDd_Q/edit?gid=1874413274#gid=1874413274).

**Comment by elena-zh** [comment]
Hey @alfetopito , I have some questions to clarify/suggestions: 
1. Why is the bff slippage added to the newly calculated one? 
image
2. Why the slippage amount to add is so huge? Is this calculated on top of the granular slippage of 20%? Buy order:
image
Sell order: 
image

3. When I keep 'Confirm' modal opened, I can see that slippage amount is constantly changing there. IMO, it is not good thing as it leads to the trade outcome change. So I'd whether show 'Price updated' warning on the modal OR freeze the slippage value on the modal until the order placement 
image
4. When a price is loaded, I can see several slippage requests on console. Is this OK? 
3 slipp

**Comment by elena-zh** [comment]
Hey @alfetopito , it works :)
But could you please explain what means 'when targeting is off' and 'fallback variation' means? :)

Also I've noticed that when:
1. 2 flags (smart slippage and fee multiplier) are off, the previously used slippage is not reset and is displayed for every trade
to 0
2. Same happens when set 'when targeting is off' to 0.

**Comment by alfetopito** [comment]
> Hey @alfetopito , I have some questions to clarify/suggestions:
> 
>     1. Why is the bff slippage added to the newly calculated one?
>        image

Already explained over the call.
In summary, for bigger amounts it'll be negligible.
For smaller amounts, BFF will be adjusted.

>     2. Why the slippage amount to add is so huge? Is this calculated on top of the granular slippage of 20%? Buy order:
>        image
>        Sell order:
>        image

Doesn't it match the calculations in the spreadsheet? I mean, apart from BFF, you can disregard that.

>     3. When I keep 'Confirm' modal opened, I can see that slippage amount is constantly changing there. IMO, it is not good thing as it leads to the trade outcome change. So I'd whether show 'Price updated' warning on the modal OR freeze the slippage value on the modal until the order placement
>        image

Good point, will fix it

>     4. When a price is loaded, I can see several slippage requests on console. Is this OK?
>        3 slipp

It's ok, as it comes from different sources

>     5. As we discussed in the meeting, we have 50% slippage limitation in the settings, so it would be nice to have it as a max when calculating a custom one
>        image

Added the limitation when both are added together to not go over 50%.

**Comment by alfetopito** [comment]
> Hey @alfetopito , it works :) But could you please explain what 'when targeting is off' and 'fallback variation' means? and dependency between them?

Targeting off => serve 0
Fallback variations => different values to serve, if we want to change to something other than 50%

> Also I've noticed that when:
> 
>     1. 2 flags (smart slippage and fee multiplier) are off, the previously used slippage is not reset and is displayed for every trade
>        to 0
> 
>     2. Same happens when set 'when targeting is off' to 0.

The slippage was not being reset when both were null and there was a previously set value.
Fixed

**Comment by alfetopito** [comment]
@elena-zh point 3 from the initial comment is fixed 🤞

File: apps/cowswap-frontend/src/modules/swap/updaters/SmartSlippageUpdater/useSmartSlippageFromFeeMultiplier.ts [hunk 8]
```diff
@@ -0,0 +1,27 @@
+import { useMemo } from 'react'
+
+import { useFeatureFlags } from '@cowprotocol/common-hooks'
+
+import { calculateBpsFromFeeMultiplier } from './calculateBpsFromFeeMultiplier'
+
+import { useReceiveAmountInfo } from '../../../trade'
```
Comment (shoom3301) [inline apps/cowswap-frontend/src/modules/swap/updaters/SmartSlippageUpdater/useSmartSlippageFromFeeMultiplier.ts:7]:
```suggestion
import { useReceiveAmountInfo } from 'modules/trade'
```

File: apps/cowswap-frontend/src/modules/swap/updaters/SmartSlippageUpdater/useSmartSlippageFromBff.ts [hunk 9]
```diff
@@ -9,9 +6,7 @@ import { useWalletInfo } from '@cowprotocol/wallet'
 import ms from 'ms.macro'
 import useSWR from 'swr'
 
-import { useDerivedTradeState, useIsWrapOrUnwrap } from 'modules/trade'
-
-import { smartSwapSlippageAtom } from '../state/slippageValueAndTypeAtom'
+import { useDerivedTradeState, useIsWrapOrUnwrap } from '../../../trade'
```
Comment (shoom3301) [inline apps/cowswap-frontend/src/modules/swap/updaters/SmartSlippageUpdater/useSmartSlippageFromBff.ts:9]:
```suggestion
import { useDerivedTradeState, useIsWrapOrUnwrap } from 'modules/trade'
```

File: apps/cowswap-frontend/src/modules/swap/updaters/SmartSlippageUpdater/useSmartSlippageFromFeeMultiplier.ts [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+
+/**
+ * Calculates smart slippage in bps, based on quoted fee
+ *
+ * Apply a multiplying factor to the fee (e.g.: 50%), and from there calculate how much slippage would be needed
+ * for the limit price to take this much more fee.
+ * More relevant for small orders in relation to fee amount, negligent for larger orders.
+ */
+export function useSmartSlippageFromFeeMultiplier(): number | undefined {
+  const { beforeNetworkCosts, afterNetworkCosts, costs, isSell } = useReceiveAmountInfo() || {}
```
Comment (shoom3301) [inline apps/cowswap-frontend/src/modules/swap/updaters/SmartSlippageUpdater/useSmartSlippageFromFeeMultiplier.ts:17]: Awesome! Thanks for the using `useReceiveAmountInfo`!

**Review by shoom3301 — APPROVED** [review]

**Comment by elena-zh** [comment]
Hey @alfetopito , all cased LGTM besides 1 case: 

> Doesn't it match the calculations in the spreadsheet? I mean, apart from BFF, you can disregard that.
For Sell orders it does not match when granular slippage is >2%:
-  500:

- 1000
sell orders

And matches when granular slippage is <2%:
match

Could you please check why?

**Comment by alfetopito** [comment]
> Hey @alfetopito , all cased LGTM besides 1 case:
> 
> > Doesn't it match the calculations in the spreadsheet? I mean, apart from BFF, you can disregard that.
> 
> For **Sell** orders it does not match when granular slippage is >2%:
> 
>     * 500:
>       
> 
>     * 1000
>       sell orders
> 
> 
> And matches when granular slippage is <2%: match
> 
> Could you please check why?
> 
> Just FYI: for Buy orders slippage values match

I believe the issues you are seeing are because the fee shown in the UI is converted to the BUY token, not the originally quoted amount, and while stablecoins are very close to 1, they are slightly off.
To be precise, you must use the quoted amounts in sell token.

**Comment by elena-zh** [comment]
Hey @alfetopito , great!
But 1 nitpick: when a final price is being loaded, it would be nice to hide the warning: 
image
as it might not be needed in the end 
image

And should we show this message to a not connected user?
image

**Comment by MindyCoW** [comment]
image

Found an issue where, even though I've set my slippage below the recommended dynamic slippage, it still shows me "High slippage amount selected".

**Comment by alfetopito** [comment]
> Hey @alfetopito , great! But 1 nitpick: when a final price is being loaded, it would be nice to hide the warning: image as it might not be needed in the end image
> 
> And should we show this message to a not connected user? image

Both fixed

**Comment by alfetopito** [comment]
> image
> 
> Found an issue where, even though I've set my slippage below the recommended dynamic slippage, it still shows me "High slippage amount selected".

That's a default message we have since the before times, for whenever the slippage is manually set to anything higher than 1%.

Would you like to change that?

**Comment by MindyCoW** [comment]
I figured, but now with the dynamic slippage it doesn't make sense to show this warning if we're recommending that the user set an even higher slippage. The warning shouldn't appear for slippage amounts equal to or lower than the recommended slippage.

BTW tooltip messages are incoming

**Review by elena-zh — APPROVED** [review]
Changes LGTM. 

Btw, I can open a separate task for changing logic for slippage warning in the settings, if needed.

**Comment by MindyCoW** [comment]
I'm going to propose a full overhaul of our slippage-related messaging. Don't get mad at me plz. 

The purpose of this is: 
- To ensure that users are aware of the new automatic adjustment of slippage
- To update some old messages to align with our current tone of voice and narrative focus
- To reflect recent understanding that ETH flow orders are visible in the public mempool and therefore not fully protected from MEV

**Swap box**

1. "Slippage tolerance" tooltip in the expandable part of the swap box: 
Pasted Graphic 20

New tooltip:
`CoW Swap dynamically adjusts your slippage tolerance to ensure your trade executes quickly while still getting the best price. Trades are protected from MEV, so your slippage can’t be exploited!`


2. High dynamic slippage warning:
Pasted Graphic 21

New text: 
`Slippage adjusted to 2.51% to ensure quick execution`


3. High dynamic slippage warning tooltip: 
Pasted Graphic 22

New tooltip: 
`CoW Swap dynamically adjusts your slippage tolerance based on current volatility. You can set a custom slippage using the gear icon above.`


4. Suggested slippage tolerance shown in swap box when a custom slippage is selected:
Pasted Graphic 25

New text: 
Instead of `Suggested: 2.33%` write `Recommended: 2.33%`


5. Suggested slippage tooltip in swap box: 
Pasted Graphic 26

New tooltip: 
`This is the recommended slippage tolerance based on current gas prices & volatility. A lower amount may result in slower execution.`


**Settings menu**

6. Slippage setting name:
image

New text: 
`MEV-protected slippage`


7. Slippage setting tooltip: 
Pasted Graphic 24

New tooltip: 
`CoW Swap dynamically adjusts your slippage tolerance to ensure your trade executes quickly while still getting the best price. To override this, enter your desired slippage amount. `

`Either way, your slippage is protected from MEV!`


8. Dynamic slippage tooltip: 
Pasted Graphic 23

New tooltip: 
`CoW Swap has dynamically selected this slippage amount to account for current gas prices and volatility. Changes may result in slower execution.`


**ETH Flow**

9. Swap box & settings menu: 
Pasted Graphic 27
Pasted Graphic 28

New tooltip: 
`When selling ETH, the minimum slippage tolerance is set to 2% to ensure a high likelihood of order execution, even in volatile market conditions.` 

`ETH orders are more vulnerable to front-running and sandwiching. For robust MEV protection, consider wrapping your ETH.
`

**Comment by alfetopito** [comment]
As Elena pointed out, makes sense to merge as is and address the texts in another PR.

That'll will come up next. Merging!

**Comment by alfetopito** [comment]
Follow up https://github.com/cowprotocol/cowswap/pull/4982

**Pull request merged**